### PR TITLE
Ports - Elise & Laneia

### PR DESCRIPTION
**Pull request by ChubbyCub**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? |Getting up to speed is going well for us because we spent time dissecting the logic and dependencies between classes. 
What _inheritance_ relations exist between classes? | CsvRecord is the abstract class that contains template methods that can be overridden by the three child classes (passenger, driver, and trip) 
What _composition_ relations exist between classes? | One passenger can have many trips, one trip can only contain one passenger/driver pair, one driver can have many trips, TripDispatcher class contains many drivers, many passengers, and many trips.
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |We decided to not raise a new ArgumentError for the case when end_time is nil in method trip_duration. The reason is the constructor does accept nil for end_time variable.
Give an example of a _template method_ that you implemented for this assignment |Method self.from_csv(record) is a template method created in CsvRecord class. This method was called in each child class. 
Give an example of a _nominal test_ that you wrote for this assignment |We wrote the test "checks to find the driver who has gone longest without trip" to test the method available_driver to make sure this method returns the driver who is available with the oldest most recent trip (or has never had a trip). 
Give an example of an _edge case test_ that you wrote for this assignment |We tested an edge case for setting driver status to a value that is neither :UNAVAILABLE nor :AVAILABLE. We expected our method to raise an argument error if an invalid status is passed into the constructor.
What is a concept that you gained more clarity on as you worked on this assignment | Gain understanding of dependencies and inheritance.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    attr_accessor :status
+
+    def initialize(id:, name:, vin:, status: :AVAILABLE, trips: nil)
+      super(id)
+
+      if vin.length != 17
+        raise ArgumentError, "VIN length is invalid"
+      end
+
+      statuses = [:AVAILABLE, :UNAVAILABLE]
```
Comment (CheezItMan) [inline lib/driver.rb:16]: This would make sense as a constant.

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    end
+
+    def average_rating
+      if @trips.length == 0
+        return 0
+      end
+
+      sum_rating = 0.00
+      num_ratings = 0
+      @trips.each do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:39]: Good work handling incomplete trips.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -27,12 +32,51 @@ def inspect
               #{passengers.count} passengers>"
     end
 
+    def available_driver
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:35]: This would make an excellent method to make private.

There is one thing, if no driver is available, you simply return `nil`, which means requesting a trip

File: lib/trip_dispatcher.rb [hunk 4]
```diff
@@ -27,12 +32,51 @@ def inspect
...
+      end
+
+      available_drivers.each do |driver|
+        if driver.trips == []
+          return driver
+        end
+      end
+      
+      available_drivers.sort_by! do |driver|
+        driver.trips.max_by { |trip| trip.end_time }.end_time
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:48]: Good use of enumerables

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -110,27 +116,72 @@
...
+        rating: nil
+      )
+      @driver.add_trip(trip1)
+      @driver.add_trip(trip2)
+    end
+
+    it "returns a float number" do
+      expect(@driver.total_revenue).must_be_kind_of Float
+    end
+
```
Comment (CheezItMan) [inline specs/driver_spec.rb:182]: You should also do a test when the trip is cheap enough that is less than or equal to 1.65, and when the trip is incomplete.

File: specs/passenger_spec.rb [hunk 6]
```diff
@@ -68,7 +76,94 @@
...
+        rating: 4
+      )
+
+      @passenger.add_trip(trip_one)
+      @passenger.add_trip(trip_two)
+      @driver.add_trip(trip_one)
+      @driver.add_trip(trip_two)
+    end
+    it "returns total costs of a passenger's trips" do
+      expect(@passenger.net_expenditures).must_equal 60
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:119]: You should also test it with incomplete trips.

File: specs/passenger_spec.rb [hunk 7]
```diff
@@ -68,7 +76,94 @@
...
+      )
+
+      @passenger.add_trip(trip_one)
+      @passenger.add_trip(trip_two)
+      @driver.add_trip(trip_one)
+      @driver.add_trip(trip_two)
+    end
+
+    it "returns total time spent on all trips" do
+      expect(@passenger.total_time_spent).must_equal 3000
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:166]: Again you should also have tests with incomplete trips.

File: specs/trip_dispatch_spec.rb [hunk 8]
```diff
@@ -122,4 +122,64 @@ def build_test_dispatcher
...
+
+    it "Checks to find the driver who has gone longest without trip" do
+      expect (@dispatcher.available_driver.id).must_equal 3
+      expect (@dispatcher.available_driver.status).must_equal :AVAILABLE
+      expect (@dispatcher.available_driver).must_be_instance_of RideShare::Driver
+    end
+
+    it "Checks passenger not assigned as driver" do
+      trip = @dispatcher.request_trip(8)
+      expect(trip.driver.id).wont_equal 8
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:162]: Technically your driver and passenger could have the same ID number.

File: specs/trip_dispatch_spec.rb [hunk 9]
```diff
@@ -122,4 +122,64 @@ def build_test_dispatcher
...
+
+    it "Check driver status is available" do
+      driver = @dispatcher.available_driver
+      expect(driver.status).must_equal :AVAILABLE
+    end
+
+    it "returns nil if all drivers are unavailable -" do
+      @dispatcher.drivers[1].status = :UNAVAILABLE
+      @dispatcher.drivers[2].status = :UNAVAILABLE
+      assert_nil(@dispatcher.available_driver)
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:173]: To do this with spec-style testing it should read:  `expect(@dispatcher.available_driver).must_be_nil`

File: specs/trip_dispatch_spec.rb [hunk 10]
```diff
@@ -122,4 +122,64 @@ def build_test_dispatcher
...
+    end
+
+    it "returns the driver who has their most recent trip ended at the oldest time" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:176]: You should also test `request_trip` when no drivers are available.

File: specs/trip_spec.rb [hunk 11]
```diff
@@ -40,4 +59,56 @@
...
+                                      status: :AVAILABLE),
+        start_time: start_time.to_s,
+        end_time: end_time.to_s,
+        cost: 23.45,
+        rating: 3,
+      }
+      @trip = RideShare::Trip.new(@trip_data)
+    end
+
+    it "returns the trip duration in seconds" do
```
Comment (CheezItMan) [inline specs/trip_spec.rb:84]: You should have a test with an incomplete trip.

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and good commit messages
Answer comprehension questions | Check, glad things got clarified
**Wave 1** |
Appropriate use of Ruby's `Time` | Check
`Trip` has a helper method to calculate duration | Check
`Passenger` methods to calculate total cost and total time | Check
Complete tests for new behavior | Check
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | Check
`Driver`s and `Trip`s are connected by `TripDispatcher` | Check
`Driver` has method to calculate average rating | Check
`Driver` has a method to calculate total revenue | Check
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check
Trip creation logic is correctly implemented | Check
New `Trip` is connected to the appropriate `Driver` and `Passenger` | Check
Tests for `request_trip` | Yes, but you needed some tests for when no drivers are available.
Wave 1 and 2 methods ignore in-progress trips |  Check
Tests for Wave 1 and 2 methods w/ in-progress trips | **NOPE**
**Wave 4 (Optional)** |
`TripDispatcher` assigns trips to the correct driver | Check
Complete tests | See my inline notes
Appropriate helper methods were made to help with complex logic | Your `available_driver` method would make an excellent private helper method.
**Overall** | Well done, you hit all the learning goals.  See my notes especially about your testing.  Let me know if you have questions, but you hit the learning goals for the project.
